src/main/java/org/elasticsearch/example/realm/CustomCachingRealm.java

```diff
@@ -64,16 +65,26 @@ public User authenticate(AuthenticationToken authenticationToken) {
         if (userHolder == null || userHolder.password == null) {
             User user = super.authenticate(token);
             if (user != null) {
-                userHolder = new UserHolder(token.credentials().copyChars(), user);
+                userHolder = new UserHolder(token.credentials().getChars(), user);
                 cache.put(token.principal(), userHolder);
                 return user;
             }
-        } else if (SecuredString.constantTimeEquals(token.credentials().internalChars(), userHolder.password)) {
+        } else if (token.credentials().equals(new SecureString(userHolder.password))) {
             return userHolder.user;
         }
         return null;
     }
 
+    /* this is the provided implementation from org.elasticsearch.xpack.security.authc.Realm */
+    @Override
+    public void authenticate(AuthenticationToken token, ActionListener<User> listener) {
```

Comment:
This is not the approach we ought to take.

The `public User authenticate(AuthenticationToken authenticationToken)` method is historical and deprecated and should no longer be used. In fact from 6.0 onwards, it doesn't even exist.
For example purposes, we should be encouraging implementors to do all their work using the async approach, and leave the blocking version as a no-op (or throw an exception).

So, essentially, pull the body of the old authenticate method here.

That will be a little weird, because there still needs to be a `User authenticate(AuthenticationToken)` method that is never called, but if anyone is building a realm for 5.5 or above, they should be preparing themselves for the 6.x API.